include/fi_log.h

```diff
@@ -35,32 +35,137 @@
 #define FI_LOG_H
 
 #if HAVE_CONFIG_H
-#  include <config.h>
+#include <config.h>
 #endif /* HAVE_CONFIG_H */
 
-extern int fi_log_level;
+#include <fi_list.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+/*
+ * Define the basic subsystems that can be used to enable fine grained logging.
+ */
+#define FABRIC (0)
+#define DOMAIN (1)
+#define EP_CM (2)
+#define EP_DM (3)
+#define AV (4)
+#define CQ (5)
+#define EQ (6)
+#define MR (7)
+
+#define LOG_WARN (1)
+#define LOG_TRACE (3)
+#define LOG_INFO (7)
+
+/*
+ * Log level occupies bit positions 0-2
+ * Subsystem occupies bit positions 3-31
+ * Provider occupes bit positions 32-63
+ */
+#define SUBSYS_OFFSET 3
+#define PROV_OFFSET 32
+
+#define MASK(width) ((UINT64_C(1) << (width)) - 1)
+
+#define SUBSYS_MASK (MASK(29) << SUBSYS_OFFSET)
+#define PROV_MASK (MASK(32) << PROV_OFFSET)
+
+/*
+ * Take bit position and offset and set bit in UINT64
+ */
+#define EXPAND(position, offset) (UINT64_C(1) << ((position) + (offset)))
```

Comment:
I would go with a single macro that takes the provider, subsystem, and level and creates log control mask from those.

I'm not sure what to call the bitmask set by the implementation, versus the mask set by the user, so I'm making stuff up now.  Maybe refer to what the code sets as the 'log_tag_bits' and keep the 'log_mask' to mean the user selected output?